infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java

```diff
@@ -258,4 +272,29 @@ private PersistentVolumeClaim replacePVCsWithCommon(
         .filter(subpath -> !isNullOrEmpty(subpath))
         .collect(Collectors.toSet());
   }
+
+  private void deleteCommonPVC(Workspace workspace) throws InfrastructureException {
+    factory.get(workspace).persistentVolumeClaims().delete(configuredPVCName);
+  }
+
+  /**
+   * @return true, if a given user has no workspaces, false otherwise
+   *
+   * @throws InfrastructureException
+   */
+  private boolean userHasNoWorkspaces() throws InfrastructureException {
+    Subject subject = EnvironmentContext.getCurrent().getSubject();
```

Comment:
This might throw an NPE in a generic case. Because `cleanup` is called just during an event handler, it might be called in a situation where there is no initiating REST call that would set the subject.

Also, generically speaking, the workspace is owned by an account, not a user, so it might belong to an organization. While we no longer support that, I think, in the UI, we should take that possibility into account (pun not intended) here.

So instead of using `EnvironmentContext` to get the current user, I would instead get the owner of the workspace using the (appropriately named) `getNamespace()` method on the workspace, check if that is indeed a user, not an organization, and then proceed from there.

Reply:
@metlos what about checking that ```EnvironmentContext.getCurrent().getSubject() == workspace.getAccount().getId()```

Reply:
> @metlos what about checking that `EnvironmentContext.getCurrent().getSubject() == workspace.getAccount().getId()`

Well, this is an event handler, so I think it is safe to assume that whatever triggered the event has checked that whoever initiated the delete has enough privs. The user might not be in the env context at all (because theoretically this can get called from a codepath not originating in a REST request)... That's why I suggested to just look at the account of the workspace, and if it is a user, check if there is no more workspaces of that user and delete the pvc.

Reply:
It will not throw NPE, but return anonymous user. I'm not exactly sure what's the consequence of that here. https://github.com/eclipse-che/che-server/blob/main/core/che-core-api-core/src/main/java/org/eclipse/che/commons/env/EnvironmentContext.java#L56

Reply:
@metlos what about 
```java
@Override
  public void cleanup(Workspace workspace) throws InfrastructureException {
    if (EphemeralWorkspaceUtility.isEphemeral(workspace)) {
      return;
    } else if (userHasNoWorkspaces(((WorkspaceImpl) workspace).getAccount())) {
      log.debug("Deleting the common PVC: '{}',", configuredPVCName);
      deleteCommonPVC(workspace);
      return;
    }
    String workspaceId = workspace.getId();
    PersistentVolumeClaim pvc = createCommonPVC(workspaceId);
    pvcSubPathHelper.removeDirsAsync(
        workspaceId,
        factory.get(workspace).getName(),
        pvc.getMetadata().getName(),
        subpathPrefixes.getWorkspaceSubPath(workspaceId));
  }
....
  private boolean userHasNoWorkspaces(AccountImpl account) throws InfrastructureException {
    try {
      if("personal".equals(account.getType())) {
        Page<WorkspaceImpl> workspaces = workspaceManager.getWorkspaces(account.getId(), false, 1, 0);
        if (workspaces.isEmpty()) {
          log.debug("User '{}' has no more workspaces left", account.getId());
          return true;
        }
      }else{
        // should never happen in 7.x since organizations is not used.
        return false;
      }
    } catch (ServerException e) {
      // should never happen
      throw new InfrastructureException(e.getLocalizedMessage(), e);
    }
    return false;
  }
```

Reply:
@skabashnyuk that is exactly what I had in mind. Thanks for stepping up and working out the details. 👍🏼

Reply:
Thanks everyone, I made some changes in the latest commit.